kafka/src/main/java/org/togglz/kafka/KafkaStateRepository.java

```diff
@@ -0,0 +1,495 @@
+      this.pollingTimeout = pollingTimeout;
+      return this;
+    }
+
+    /**
+     * Defines an upper bound for the initialization time of the {@link KafkaStateRepository}.
+     * If the underlying Kafka consumer is not able to read all data from Kafka within the
+     * given interval, the {@link #build()} method will throw an exception.
+     *
+     * @param initializationTimeout the upper bound for the initialization time
+     * @return the current instance of this builder
+     */
+    public Builder initializationTimeout(Duration initializationTimeout) {
+      this.initializationTimeout = initializationTimeout;
+      return this;
+    }
+
+    /**
+     * Returns a new {@link KafkaStateRepository} instance.
+     *
+     * @return the new {@link KafkaStateRepository} instance
+     * @throws NullPointerException if any of the configuration values is {@code null}
+     * @throws RuntimeException if the {@link KafkaStateRepository} can't be initialized
+     *         within the defined interval
+     */
+    public KafkaStateRepository build() {
+      return new KafkaStateRepository(this);
+    }
+
+  }
+
+  static class FeatureStateConsumer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FeatureStateConsumer.class);
+    private static final Gson GSON = new GsonBuilder().create();
+
+    private final KafkaConsumer<String, String> kafkaConsumer;
+    private final BiConsumer<String, FeatureStateStorageWrapper> updateHandler;
+    private final String inboundTopic;
+    private final Duration pollingTimeout;
+    private final Map<TopicPartition, Long> offsets;
+    private final CountDownLatch initializationLatch;
+    private final Duration initializationTimeout;
+    private final CountDownLatch shutdownLatch;
+
+    private volatile boolean running;
+    private volatile long consumerLag;
+
+    FeatureStateConsumer(Builder builder, BiConsumer<String, FeatureStateStorageWrapper> updateHandler) {
+      Properties properties = new Properties();
+      properties.setProperty(BOOTSTRAP_SERVERS_CONFIG, requireNonNull(builder.bootstrapServers));
+      properties.setProperty(KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+      properties.setProperty(VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+      properties.setProperty(ENABLE_AUTO_COMMIT_CONFIG, "false");
+
+      this.kafkaConsumer = new KafkaConsumer<>(properties);
+      this.updateHandler = updateHandler;
+      this.inboundTopic = requireNonNull(builder.inboundTopic);
+      this.pollingTimeout = requireNonNull(builder.pollingTimeout);
+      this.offsets = new ConcurrentHashMap<>();
+      this.initializationLatch = new CountDownLatch(1);
+      this.initializationTimeout = requireNonNull(builder.initializationTimeout);
+      this.shutdownLatch = new CountDownLatch(1);
+
+      running = false;
+      consumerLag = Long.MAX_VALUE;
+    }
+
+    void start() {
+      if (running) {
```

Comment:
Good point. I will remove the 'double-check-locking' and synchronize the start/stop method. But I will keep the volatile declaration for running/consumerLag (since I don't want to synchronize the accessor methods for those fields).